File: spring-web-java/src/main/java/spring/web/java/presentation/controller/OrderController.java
package spring.web.java.presentation.controller;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import spring.web.java.application.service.OrderService;
import spring.web.java.domain.model.enums.DeliveryStatus;
import spring.web.java.domain.model.enums.OrderStatus;
import spring.web.java.infrastructure.aspect.RequestLock;
import spring.web.java.presentation.dto.response.OrderResponse;
import spring.web.java.presentation.dto.request.OrderSaveRequest;

@RestController
@RequestMapping("/orders")
@RequiredArgsConstructor
public class OrderController {

	private final OrderService orderService;

	@RequestLock
	@PreAuthorize("isAuthenticated()")
	@PostMapping
	@ResponseStatus(HttpStatus.CREATED)
	public OrderResponse saveOrder(@RequestBody @Valid OrderSaveRequest orderSaveRequest) {
		return orderService.saveOrder(orderSaveRequest);
	}

	@PreAuthorize("isAuthenticated()")
	@GetMapping
	public Page<OrderResponse> findOrders(
		@PageableDefault Pageable pageable,
		@RequestParam Long memberId,
		@RequestParam(required = false) OrderStatus orderStatus,
		@RequestParam(required = false) DeliveryStatus deliveryStatus
	) {
		return orderService.findOrders(memberId, orderStatus, deliveryStatus, pageable);
	}

	@PreAuthorize("isAuthenticated()")
	@GetMapping("/{id}")
	public OrderResponse findOrder(@PathVariable Long id) {
		return orderService.findOrder(id);
	}

	@RequestLock
	@PreAuthorize("isAuthenticated()")
	@PostMapping("/{id}")
	public OrderResponse cancelOrder(@PathVariable Long id) {
		return orderService.cancelOrder(id);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/presentation/controller/OrderController.kt
package spring.web.kotlin.presentation.controller

import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.HttpStatus
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import spring.web.kotlin.application.service.OrderService
import spring.web.kotlin.domain.model.enums.DeliveryStatus
import spring.web.kotlin.domain.model.enums.OrderStatus
import spring.web.kotlin.infrastructure.aspect.RequestLock
import spring.web.kotlin.presentation.dto.request.OrderSaveRequest

@RestController
@RequestMapping("/orders")
class OrderController(
    private val orderService: OrderService
) {
    @PostMapping
    @PreAuthorize("isAuthenticated()")
    @RequestLock
    @ResponseStatus(HttpStatus.CREATED)
    fun saveOrder(@RequestBody @Validated orderSaveRequest: OrderSaveRequest) = orderService.saveOrder(orderSaveRequest)

    @GetMapping
    @PreAuthorize("isAuthenticated()")
    fun findOrders(
        @PageableDefault pageable: Pageable,
        @RequestParam(required = false) memberId: Long?,
        @RequestParam(required = false) orderStatus: OrderStatus?,
        @RequestParam(required = false) deliveryStatus: DeliveryStatus?,
    ) = orderService.findOrders(memberId, orderStatus, deliveryStatus, pageable)

    @GetMapping("/{id}")
    @PreAuthorize("isAuthenticated()")
    fun findOrder(@PathVariable id: Long) = orderService.findOrder(id)

    @PostMapping("/{id}")
    @PreAuthorize("isAuthenticated()")
    @RequestLock
    fun cancelOrder(@PathVariable id: Long) = orderService.cancelOrder(id)
}